### ISSUE TITLE
Installing dev_environment.sh fails

### ISSUE THREAD

**Issue by trhyder** [IfcOpenShell/IfcOpenShell#5475]
I get the below error when trying to run dev_environment.sh

It could be a syntax issue with how spaces are handled.

This may not be the correct process for contributing to the repo through PRs, but the docs don't cover this explicitly. Here is what I did:
- I successfully installed a daily build of Bonsai
![image](https://github.com/user-attachments/assets/01f4b0c9-8b95-4599-933f-f71ddfb115fb)
- I forked a repo to my Git account
- I cloned my forked repo to my computer
- I made a new branch to address a specific issue
- I then ran from terminal `./dev_environment.sh`

The third-party dependencies seem to have been installed to my repo location, but it seems not to be finding my blender path

```
tim@Tims-Mac-Studio installation % ./dev_environment.sh                                                                     
cp: /Users/tim/Library/Application: No such file or directory
cp: Support/Blender/4.2/extensions/.local/lib/python3.11/site-packages/ifcopenshell/*_wrapper*: No such file or directory
rm: /Users/tim/Library/Application: No such file or directory
rm: Support/Blender/4.2/extensions/user_default/bonsai/__init__.py: No such file or directory
rm: /Users/tim/Library/Application: No such file or directory
rm: Support/Blender/4.2/extensions/.local/lib/python3.11/site-packages/bonsai: No such file or directory
rm: /Users/tim/Library/Application: No such file or directory
rm: Support/Blender/4.2/extensions/.local/lib/python3.11/site-packages/ifcopenshell: No such file or directory
rm: /Users/tim/Library/Application: No such file or directory
rm: Support/Blender/4.2/extensions/.local/lib/python3.11/site-packages/ifccsv.py: No such file or directory
rm: /Users/tim/Library/Application: No such file or directory
rm: Support/Blender/4.2/extensions/.local/lib/python3.11/site-packages/ifcdiff.py: No such file or directory
rm: /Users/tim/Library/Application: No such file or directory
rm: Support/Blender/4.2/extensions/.local/lib/python3.11/site-packages/bsdd.py: No such file or directory
rm: /Users/tim/Library/Application: No such file or directory
rm: Support/Blender/4.2/extensions/.local/lib/python3.11/site-packages/bcf: No such file or directory
rm: /Users/tim/Library/Application: No such file or directory
rm: Support/Blender/4.2/extensions/.local/lib/python3.11/site-packages/ifc4d: No such file or directory
rm: /Users/tim/Library/Application: No such file or directory
rm: Support/Blender/4.2/extensions/.local/lib/python3.11/site-packages/ifc5d: No such file or directory
rm: /Users/tim/Library/Application: No such file or directory
rm: Support/Blender/4.2/extensions/.local/lib/python3.11/site-packages/ifccityjson: No such file or directory
rm: /Users/tim/Library/Application: No such file or directory
rm: Support/Blender/4.2/extensions/.local/lib/python3.11/site-packages/ifcclash: No such file or directory
rm: /Users/tim/Library/Application: No such file or directory
rm: Support/Blender/4.2/extensions/.local/lib/python3.11/site-packages/ifcpatch: No such file or directory
rm: /Users/tim/Library/Application: No such file or directory
rm: Support/Blender/4.2/extensions/.local/lib/python3.11/site-packages/ifctester: No such file or directory
rm: /Users/tim/Library/Application: No such file or directory
rm: Support/Blender/4.2/extensions/.local/lib/python3.11/site-packages/ifcfm: No such file or directory
ln: Support/Blender/4.2/extensions/user_default/bonsai/__init__.py: No such file or directory
ln: Support/Blender/4.2/extensions/.local/lib/python3.11/site-packages/bonsai: No such file or directory
ln: Support/Blender/4.2/extensions/.local/lib/python3.11/site-packages/ifcopenshell: No such file or directory
ln: Support/Blender/4.2/extensions/.local/lib/python3.11/site-packages/ifccsv.py: No such file or directory
ln: Support/Blender/4.2/extensions/.local/lib/python3.11/site-packages/ifcdiff.py: No such file or directory
ln: Support/Blender/4.2/extensions/.local/lib/python3.11/site-packages/bsdd.py: No such file or directory
ln: Support/Blender/4.2/extensions/.local/lib/python3.11/site-packages/bcf: No such file or directory
ln: Support/Blender/4.2/extensions/.local/lib/python3.11/site-packages/ifc4d: No such file or directory
ln: Support/Blender/4.2/extensions/.local/lib/python3.11/site-packages/ifc5d: No such file or directory
ln: Support/Blender/4.2/extensions/.local/lib/python3.11/site-packages/ifccityjson: No such file or directory
ln: Support/Blender/4.2/extensions/.local/lib/python3.11/site-packages/ifcclash: No such file or directory
ln: Support/Blender/4.2/extensions/.local/lib/python3.11/site-packages/ifcpatch: No such file or directory
ln: Support/Blender/4.2/extensions/.local/lib/python3.11/site-packages/ifctester: No such file or directory
ln: Support/Blender/4.2/extensions/.local/lib/python3.11/site-packages/ifcfm: No such file or directory
./dev_environment.sh: line 48: cd: /Users/tim/Library/Application: No such file or directory
--2024-09-24 09:15:28--  https://raw.githubusercontent.com/jsGanttImproved/jsgantt-improved/master/dist/jsgantt.js
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.110.133, 185.199.111.133, 185.199.108.133, ...
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.110.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 233737 (228K) [text/plain]
Saving to: ‘jsgantt.js’

jsgantt.js                                                                 100%[======================================================================================================================================================================================>] 228.26K  --.-KB/s    in 0.02s   

2024-09-24 09:15:28 (11.2 MB/s) - ‘jsgantt.js’ saved [233737/233737]

--2024-09-24 09:15:28--  https://raw.githubusercontent.com/jsGanttImproved/jsgantt-improved/master/dist/jsgantt.css
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.110.133, 185.199.111.133, 185.199.108.133, ...
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.110.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 18913 (18K) [text/plain]
Saving to: ‘jsgantt.css’

jsgantt.css                                                                100%[======================================================================================================================================================================================>]  18.47K  --.-KB/s    in 0.002s  

2024-09-24 09:15:28 (8.03 MB/s) - ‘jsgantt.css’ saved [18913/18913]

./dev_environment.sh: line 51: cd: /Users/tim/Library/Application: No such file or directory
--2024-09-24 09:15:28--  https://github.com/BrickSchema/Brick/releases/download/nightly/Brick.ttl
Resolving github.com (github.com)... 140.82.112.4
Connecting to github.com (github.com)|140.82.112.4|:443... connected.
HTTP request sent, awaiting response... 302 Found
Location: https://objects.githubusercontent.com/github-production-release-asset-2e65be/67053564/038242f9-b0ba-4bf8-a521-d882ea38e7d5?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=releaseassetproduction%2F20240924%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20240924T131528Z&X-Amz-Expires=300&X-Amz-Signature=2a53fd97b86c8c4e2ad46cc95ec3e5beaa9be6ae64f3531379480363c3baf8e9&X-Amz-SignedHeaders=host&response-content-disposition=attachment%3B%20filename%3DBrick.ttl&response-content-type=application%2Foctet-stream [following]
--2024-09-24 09:15:28--  https://objects.githubusercontent.com/github-production-release-asset-2e65be/67053564/038242f9-b0ba-4bf8-a521-d882ea38e7d5?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=releaseassetproduction%2F20240924%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20240924T131528Z&X-Amz-Expires=300&X-Amz-Signature=2a53fd97b86c8c4e2ad46cc95ec3e5beaa9be6ae64f3531379480363c3baf8e9&X-Amz-SignedHeaders=host&response-content-disposition=attachment%3B%20filename%3DBrick.ttl&response-content-type=application%2Foctet-stream
Resolving objects.githubusercontent.com (objects.githubusercontent.com)... 185.199.110.133, 185.199.108.133, 185.199.111.133, ...
Connecting to objects.githubusercontent.com (objects.githubusercontent.com)|185.199.110.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 1760685 (1.7M) [application/octet-stream]
Saving to: ‘Brick.ttl’

Brick.ttl                                                                  100%[======================================================================================================================================================================================>]   1.68M  --.-KB/s    in 0.1s    

2024-09-24 09:15:29 (13.5 MB/s) - ‘Brick.ttl’ saved [1760685/1760685]
```

**Comment by Moult** [comment]
How have you defined your variables at the top? Are the paths quoted?

**Comment by trhyder** [comment]
copied and pasted

```
REPO_PATH="/Users/tim/repos/IfcOpenShell"
BLENDER_PATH="/Users/tim/Library/Application Support/Blender/4.2"
PACKAGE_PATH="${BLENDER_PATH}/extensions/.local/lib/python3.11/site-packages"
BONSAI_PATH="${BLENDER_PATH}/extensions/user_default/bonsai"
```

**Comment by trhyder** [comment]
@Moult thanks for your help - I am still getting some errors though

should `BONSAI_PATH="${BLENDER_PATH}/extensions/raw_githubusercontent_com/bonsai"` rather than `BONSAI_PATH="${BLENDER_PATH}/extensions/user_default/bonsai"`?

and this is from my site_packages folder

<img width="341" alt="image" src="https://github.com/user-attachments/assets/fd97752f-be91-4e64-b4ca-f02ed757e215">


```
tim@Tims-Mac-Studio installation % ./dev_environment.sh
cp: /Users/tim/Library/Application Support/Blender/4.2/extensions/.local/lib/python3.11/site-packages/ifcopenshell/*_wrapper*: No such file or directory
rm: /Users/tim/Library/Application Support/Blender/4.2/extensions/user_default/bonsai/__init__.py: No such file or directory
rm: /Users/tim/Library/Application Support/Blender/4.2/extensions/.local/lib/python3.11/site-packages/bonsai: No such file or directory
ln: /Users/tim/Library/Application Support/Blender/4.2/extensions/user_default/bonsai/__init__.py: No such file or directory
ln: $PACKAGE_PATH/bonsai: No such file or directory
./dev_environment.sh: line 49: cd: /Users/tim/Library/Application Support/Blender/4.2/extensions/.local/lib/python3.11/site-packages/bonsai/bim/data/gantt: No such file or directory
--2024-09-24 21:09:24--  https://raw.githubusercontent.com/jsGanttImproved/jsgantt-improved/master/dist/jsgantt.js
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.111.133, 185.199.110.133, 185.199.109.133, ...
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.111.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 233737 (228K) [text/plain]
Saving to: ‘jsgantt.js.3’

jsgantt.js.3                                                               100%[=======================================================================================================================================================================================>] 228.26K  --.-KB/s    in 0.02s   

2024-09-24 21:09:24 (10.1 MB/s) - ‘jsgantt.js.3’ saved [233737/233737]

--2024-09-24 21:09:24--  https://raw.githubusercontent.com/jsGanttImproved/jsgantt-improved/master/dist/jsgantt.css
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.111.133, 185.199.110.133, 185.199.109.133, ...
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.111.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 18913 (18K) [text/plain]
Saving to: ‘jsgantt.css.3’

jsgantt.css.3                                                              100%[=======================================================================================================================================================================================>]  18.47K  --.-KB/s    in 0.001s  

2024-09-24 21:09:24 (12.3 MB/s) - ‘jsgantt.css.3’ saved [18913/18913]

./dev_environment.sh: line 52: cd: /Users/tim/Library/Application Support/Blender/4.2/extensions/.local/lib/python3.11/site-packages/bonsai/bim/schema: No such file or directory
--2024-09-24 21:09:24--  https://github.com/BrickSchema/Brick/releases/download/nightly/Brick.ttl
Resolving github.com (github.com)... 140.82.113.3
Connecting to github.com (github.com)|140.82.113.3|:443... connected.
HTTP request sent, awaiting response... 302 Found
Location: https://objects.githubusercontent.com/github-production-release-asset-2e65be/67053564/038242f9-b0ba-4bf8-a521-d882ea38e7d5?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=releaseassetproduction%2F20240925%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20240925T010924Z&X-Amz-Expires=300&X-Amz-Signature=b89287b8a56ad7c74b218717ee414b91999c1a199eb022f6692c2a0d68455e20&X-Amz-SignedHeaders=host&response-content-disposition=attachment%3B%20filename%3DBrick.ttl&response-content-type=application%2Foctet-stream [following]
--2024-09-24 21:09:24--  https://objects.githubusercontent.com/github-production-release-asset-2e65be/67053564/038242f9-b0ba-4bf8-a521-d882ea38e7d5?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=releaseassetproduction%2F20240925%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20240925T010924Z&X-Amz-Expires=300&X-Amz-Signature=b89287b8a56ad7c74b218717ee414b91999c1a199eb022f6692c2a0d68455e20&X-Amz-SignedHeaders=host&response-content-disposition=attachment%3B%20filename%3DBrick.ttl&response-content-type=application%2Foctet-stream
Resolving objects.githubusercontent.com (objects.githubusercontent.com)... 185.199.109.133, 185.199.110.133, 185.199.108.133, ...
Connecting to objects.githubusercontent.com (objects.githubusercontent.com)|185.199.109.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 1760685 (1.7M) [application/octet-stream]
Saving to: ‘Brick.ttl.3’

Brick.ttl.3                                                                100%[=======================================================================================================================================================================================>]   1.68M  10.1MB/s    in 0.2s    

2024-09-24 21:09:24 (10.1 MB/s) - ‘Brick.ttl.3’ saved [1760685/1760685]
```

**Comment by Moult** [comment]
> should BONSAI_PATH="${BLENDER_PATH}/extensions/raw_githubusercontent_com/bonsai"

Yes that's a better default. Fixed.

I also made a mistake in quotes, fixed now.

Also you might want to start fresh as likely running the script previously borked things so even with a fixed script it still won't work.

**Comment by trhyder** [comment]
Hey @Moult, it's still not quite working for me. I get the below terminal error:

```
tim@Tims-Mac-Studio installation % ./dev_environment.sh                                                                                    
cp: /Users/tim/Library/Application Support/Blender/4.2/extensions/.local/lib/python3.11/site-packages/ifcopenshell/*_wrapper*: No such file or directory
--2024-09-25 14:13:31--  https://raw.githubusercontent.com/jsGanttImproved/jsgantt-improved/master/dist/jsgantt.js
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.109.133, 185.199.111.133, 185.199.108.133, ...
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.109.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 233737 (228K) [text/plain]
Saving to: ‘jsgantt.js.3’

jsgantt.js.3                                                               100%[=======================================================================================================================================================================================>] 228.26K  --.-KB/s    in 0.02s   

2024-09-25 14:13:31 (11.3 MB/s) - ‘jsgantt.js.3’ saved [233737/233737]

--2024-09-25 14:13:31--  https://raw.githubusercontent.com/jsGanttImproved/jsgantt-improved/master/dist/jsgantt.css
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.109.133, 185.199.111.133, 185.199.108.133, ...
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.109.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 18913 (18K) [text/plain]
Saving to: ‘jsgantt.css.3’

jsgantt.css.3                                                              100%[=======================================================================================================================================================================================>]  18.47K  --.-KB/s    in 0.001s  

2024-09-25 14:13:31 (18.2 MB/s) - ‘jsgantt.css.3’ saved [18913/18913]

--2024-09-25 14:13:31--  https://github.com/BrickSchema/Brick/releases/download/nightly/Brick.ttl
Resolving github.com (github.com)... 140.82.113.3
Connecting to github.com (github.com)|140.82.113.3|:443... connected.
HTTP request sent, awaiting response... 302 Found
Location: https://objects.githubusercontent.com/github-production-release-asset-2e65be/67053564/538421d7-b38a-49f2-8824-c8edcddf6496?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=releaseassetproduction%2F20240925%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20240925T181331Z&X-Amz-Expires=300&X-Amz-Signature=ce1aa6f03ea8d16694e4bd00c4ae7e3c7135b43f1df5c9ec7b6daf0923b6ad97&X-Amz-SignedHeaders=host&response-content-disposition=attachment%3B%20filename%3DBrick.ttl&response-content-type=application%2Foctet-stream [following]
--2024-09-25 14:13:31--  https://objects.githubusercontent.com/github-production-release-asset-2e65be/67053564/538421d7-b38a-49f2-8824-c8edcddf6496?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=releaseassetproduction%2F20240925%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20240925T181331Z&X-Amz-Expires=300&X-Amz-Signature=ce1aa6f03ea8d16694e4bd00c4ae7e3c7135b43f1df5c9ec7b6daf0923b6ad97&X-Amz-SignedHeaders=host&response-content-disposition=attachment%3B%20filename%3DBrick.ttl&response-content-type=application%2Foctet-stream
Resolving objects.githubusercontent.com (objects.githubusercontent.com)... 185.199.109.133, 185.199.108.133, 185.199.110.133, ...
Connecting to objects.githubusercontent.com (objects.githubusercontent.com)|185.199.109.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 1760685 (1.7M) [application/octet-stream]
Saving to: ‘Brick.ttl.2’

Brick.ttl.2                                                                100%[=======================================================================================================================================================================================>]   1.68M  --.-KB/s    in 0.1s    

2024-09-25 14:13:31 (11.6 MB/s) - ‘Brick.ttl.2’ saved [1760685/1760685]
```

**Comment by Moult** [comment]
```
cp: /Users/tim/Library/Application Support/Blender/4.2/extensions/.local/lib/python3.11/site-packages/ifcopenshell/*_wrapper*: No such file or directory
```

That's the key error here. In a fresh install prior to running the dev script can you show what the contents of `/Users/tim/Library/Application Support/Blender/4.2/extensions/.local/lib/python3.11/site-packages/ifcopenshell/*_wrapper*` is?

**Comment by trhyder** [comment]
@Moult, thanks for your persistence. Sorry, I don't understand what `/*_wrapper*` is meant to show. I get a "No Such file or directory" when I include it, if I drop it off I do get a list. Below are the commands I entered into terminal and the responses:

```
tim@Tims-Mac-Studio ~ % ls "/Users/tim/Library/Application Support/Blender/4.2/extensions/.local/lib/python3.11/site-packages/ifcopenshell/*_wrapper*"
ls: /Users/tim/Library/Application Support/Blender/4.2/extensions/.local/lib/python3.11/site-packages/ifcopenshell/*_wrapper*: No such file or directory
tim@Tims-Mac-Studio ~ % ls "/Users/tim/Library/Application Support/Blender/4.2/extensions/.local/lib/python3.11/site-packages/ifcopenshell"          
__init__.py		entity_instance.py	ifcopenshell_wrapper.py	sql.py			validate.py
__pycache__		express			mvd			stream.py
alignment.py		file.py			py.typed		template.py
api			geom			requirements.txt	transition_curve.py
draw.py			guid.py			settings.py		util
```

**Comment by Moult** [comment]
There are two issues here. First is my bad, I learned today that MacOS wildcards in quotes mean something else: https://superuser.com/questions/173287/cp-command-not-working-with-wildcards-on-mac-os-x

So I guess the command needs to change to this in the script (notice the subtle relocation of the quote):

```
cp "${PACKAGE_PATH}/ifcopenshell/"*_wrapper* "${PWD}/src/ifcopenshell-python/ifcopenshell/"
```

The second issue is that you seem to be missing a critical file. The two files you should ahve in that folder are:

 - ifcopenshell_wrapper.py (you have this, good)
 - _ifcopenshell_wrapper.cpython-311-darwin.so (missing)

I think it's missing because our builds were recently broken for MacOS and @Andrej730 has [fixed it](https://github.com/IfcOpenShell/IfcOpenShell/commit/700a453501dc10f1fe4b17c70581fe984ac22d34) yesterday. Can you try:

 1. Reinstall the latest unstable from a fresh Blender 4.2 and confirm that the missing .so file is now present
 2. Try the script again with that line modified with the new quote location.

If that works we can update the script.

**Comment by trhyder** [comment]
Thanks for the help it worked!

1. I removed my install of Blender, deleted `/Users/tim/Library/Application Support/Blender/4.2`, installed a fresh version of  Blender 4.2.2, and installed Bonsai through the repository `https://raw.githubusercontent.com/IfcOpenShell/bonsai_unstable_repo/main/index.json`

```
tim@Tims-Mac-Studio ~ % ls "/Users/tim/Library/Application Support/Blender/4.2/extensions/.local/lib/python3.11/site-packages/ifcopenshell/"*_wrapper*
/Users/tim/Library/Application Support/Blender/4.2/extensions/.local/lib/python3.11/site-packages/ifcopenshell/_ifcopenshell_wrapper.cpython-311-darwin.so
/Users/tim/Library/Application Support/Blender/4.2/extensions/.local/lib/python3.11/site-packages/ifcopenshell/ifcopenshell_wrapper.py
```

2. I synced my fork with IfcOpenShell on github, edited `dev_environment.sh` to point to my local directors and moved the  wildcard out of the quotes. Running the script successfully redirected to my local repo

One question I still have is whether there should be a section in the documentation explaining how to contribute to the repo and keep up-to-date with the latest. This may be obvious to experienced developers, but I am still not sure I am doing it the best way.
